File: sdk/python/nova_core_sdk/streaming.py
from __future__ import annotations

import json
from typing import Any, AsyncIterator, Iterator


class StreamProcessor:
    def __init__(self, response: Any) -> None:
        self._response = response

    def __iter__(self) -> Iterator[dict[str, Any]]:
        for line in self._response.iter_lines():
            if not line:
                continue
            line_str = line if isinstance(line, str) else line.decode("utf-8")
            if line_str.startswith("data: "):
                data_str = line_str[6:]
                if data_str.strip() == "[DONE]":
                    break
                try:
                    yield json.loads(data_str)
                except json.JSONDecodeError:
                    continue

    def text(self) -> str:
        chunks: list[str] = []
        for chunk in self:
            if "choices" in chunk:
                for choice in chunk["choices"]:
                    delta = choice.get("delta", {})
                    content = delta.get("content", "")
                    if content:
                        chunks.append(content)
            elif "content" in chunk:
                chunks.append(chunk["content"])
        return "".join(chunks)


class AsyncStreamProcessor:
    def __init__(self, response: Any) -> None:
        self._response = response

    async def __aiter__(self) -> AsyncIterator[dict[str, Any]]:
        async for line in self._response.aiter_lines():
            if not line:
                continue
            if line.startswith("data: "):
                data_str = line[6:]
                if data_str.strip() == "[DONE]":
                    break
                try:
                    yield json.loads(data_str)
                except json.JSONDecodeError:
                    continue

    async def text(self) -> str:
        chunks: list[str] = []
        async for chunk in self:
            if "choices" in chunk:
                for choice in chunk["choices"]:
                    delta = choice.get("delta", {})
                    content = delta.get("content", "")
                    if content:
                        chunks.append(content)
            elif "content" in chunk:
                chunks.append(chunk["content"])
        return "".join(chunks)

File: sdk/typescript/src/streaming.ts
import { Configuration, getDefaultHeaders } from "./config.js";
import type { StreamChunk } from "./types.js";

export class StreamProcessor {
  private reader: ReadableStreamDefaultReader<Uint8Array> | null = null;
  private buffer: string = "";

  constructor(private response: Response) {
    if (response.body) {
      this.reader = response.body.getReader();
    }
  }

  async *iter(): AsyncGenerator<StreamChunk> {
    if (!this.reader) return;

    while (true) {
      const { done, value } = await this.reader.read();
      if (done) break;

      this.buffer += new TextDecoder().decode(value);
      const lines = this.buffer.split("\n");
      this.buffer = lines.pop() || "";

      for (const line of lines) {
        const trimmed = line.trim();
        if (!trimmed || !trimmed.startsWith("data: ")) continue;
        const data = trimmed.slice(6);
        if (data === "[DONE]") return;
        try {
          yield JSON.parse(data) as StreamChunk;
        } catch {
          continue;
        }
      }
    }
  }

  async text(): Promise<string> {
    const chunks: string[] = [];
    for await (const chunk of this.iter()) {
      if (chunk.choices) {
        for (const choice of chunk.choices) {
          const content = choice.delta?.content || choice.text || "";
          if (content) chunks.push(content);
        }
      } else if (chunk.content) {
        chunks.push(chunk.content);
      }
    }
    return chunks.join("");
  }
}

export async function createStream(
  config: Configuration,
  path: string,
  body: unknown,
): Promise<StreamProcessor> {
  const url = new URL(path, config.baseUrl);
  const headers = {
    ...getDefaultHeaders(config),
    Accept: "text/event-stream",
  };

  const response = await fetch(url.toString(), {
    method: "POST",
    headers,
    body: JSON.stringify(body),
  });

  return new StreamProcessor(response);
}
